policy-controller/k8s/status/src/index.rs

```diff
@@ -0,0 +1,246 @@
+                        observed_generation: None,
+                        reason: "NoMatchingParent".to_string(),
+                        status: "False".to_string(),
+                        type_: "Accepted".to_string(),
+                    }
+                };
+                gateway::RouteParentStatus {
+                    parent_ref: gateway::ParentReference {
+                        group: Some(POLICY_API_GROUP.to_string()),
+                        kind: Some("Server".to_string()),
+                        namespace: Some(parent_reference_id.namespace.clone()),
+                        name: parent_reference_id.name.clone(),
+                        section_name: None,
+                        port: None,
+                    },
+                    controller_name: STATUS_CONTROLLER_NAME.to_string(),
+                    conditions: vec![condition],
+                }
+            })
+            .collect();
+        let status = gateway::HttpRouteStatus {
+            inner: gateway::RouteStatus {
+                parents: parent_statuses,
+            },
+        };
+        let patch = serde_json::json!({
+            "apiVersion": POLICY_API_VERSION,
+            "kind": "HTTPRoute",
+            "name": route_id.name,
+            "status": status,
+        });
+
+        // Patch the HTTPRoute with its status.
+        let api = k8s::Api::<k8s::policy::HttpRoute>::namespaced(
+            self.client.clone(),
+            &route_id.namespace,
+        );
+        if let Err(error) = api
+            .patch_status(&route_id.name, &patch_params, &k8s::Patch::Merge(patch))
+            .await
+        {
+            tracing::error!(namespace = %route_id.namespace, name = %route_id.name, %error, "Failed to patch HTTPRoute");
+        }
+    }
+
+    async fn process_server_update(&mut self, patch_params: k8s::PatchParams) {
+        let route_ids: Vec<ResourceId> = self.index.read().http_routes.keys().cloned().collect();
+        for route_id in route_ids {
+            self.process_http_route_update(route_id, patch_params.clone())
+                .await;
+        }
+    }
+}
+
+impl Index {
+    pub fn shared(updates: UnboundedSender<Update>) -> SharedIndex {
+        Arc::new(RwLock::new(Self {
+            updates,
+            http_routes: HashMap::new(),
+            servers: HashSet::new(),
+        }))
+    }
+
+    fn update_http_route(&mut self, route_id: ResourceId, route_binding: RouteBinding) -> bool {
+        match self.http_routes.entry(route_id) {
+            Entry::Vacant(entry) => {
+                entry.insert(route_binding);
+            }
+            Entry::Occupied(mut entry) => {
+                if *entry.get() == route_binding {
+                    return false;
+                }
+                entry.insert(route_binding);
+            }
+        }
+        true
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::policy::HttpRoute> for Index {
```

Comment:
should we `use` instead of namespace/qualify the struct here?

Reply:
We do the same in policy controller's index code. We could use a `use` to shorten this, but we don't use `kubert` anywhere else in the file so I kept it the same.